File: examples/dekita/.claude/hooks/ui-check-reminder.py
#!/usr/bin/env python3
"""フロントエンド変更時にブラウザ確認を強制。

Why:
    フロントエンドファイル変更後にブラウザ確認せずコミットすると、
    ランタイムエラーやUI崩れに気づかないまま本番に反映される。

What:
    - git commit時（PreToolUse:Bash）に発火
    - frontend/src/配下のTS/TSX/JSON/CSS変更を検出
    - ブラウザ確認マーカーファイルがなければブロック
    - confirm-ui-check.py実行後にコミット可能

State:
    - reads: .claude/logs/markers/ui-check-*.done

Remarks:
    - ブロック型フック（確認なしはコミット不可）
    - main/masterブランチはスキップ
    - .ts/.tsxはlib/hooks/workersも含む（Issue #209）

Changelog:
    - silenvx/dekita#xxx: フック追加
    - silenvx/dekita#209: .ts/.tsxファイル対象を拡大
"""

import fnmatch
import json
import re
import subprocess
import sys

from common import MARKERS_LOG_DIR
from lib.constants import TIMEOUT_MEDIUM
from lib.execution import log_hook_execution
from lib.git import get_current_branch
from lib.results import make_block_result
from lib.session import parse_hook_input
from lib.strings import sanitize_branch_name, strip_quoted_strings

# File patterns that require browser verification
# These patterns are matched against git paths (relative to repo root)
# Expanded to cover ALL frontend source files (see Issue #209 for background)
FRONTEND_FILE_PATTERNS = [
    "frontend/src/**/*.ts",  # All TypeScript files (lib, hooks, workers, etc.)
    "frontend/src/**/*.tsx",  # All React components
    "frontend/src/i18n/locales/*.json",  # i18n translations
    "frontend/src/index.css",  # Global CSS
]


def matches_frontend_pattern(filepath: str) -> bool:
    """Check if a file path matches any frontend file pattern."""
    for pattern in FRONTEND_FILE_PATTERNS:
        # fnmatch doesn't support **, so we need to handle it manually
        if "**" in pattern:
            # Split pattern at **
            parts = pattern.split("**")
            if len(parts) == 2:
                prefix, suffix = parts
                # Check if file starts with prefix and ends with suffix pattern
                if filepath.startswith(prefix):
                    remaining = filepath[len(prefix) :]
                    # suffix might have a leading /, remove it
                    if suffix.startswith("/"):
                        suffix = suffix[1:]
                    if fnmatch.fnmatch(remaining, "*" + suffix) or fnmatch.fnmatch(
                        remaining, "*/" + suffix
                    ):
                        return True
        else:
            if fnmatch.fnmatch(filepath, pattern):
                return True
    return False


def is_git_commit_command(command: str) -> bool:
    """Check if command is a git commit command."""
    if not command.strip():
        return False

    stripped_command = strip_quoted_strings(command)
    return bool(re.search(r"git\s+commit\b", stripped_command))


def has_auto_stage_flag(command: str) -> bool:
    """Check if command has -a or -am flag (auto-staging modified files)."""
    stripped_command = strip_quoted_strings(command)
    # Match -a, -am, -ma, or --all flags
    return bool(re.search(r"git\s+commit\s+.*(-a\b|--all\b|-[a-z]*a[a-z]*\b)", stripped_command))


def get_staged_frontend_files() -> list[str]:
    """Get list of staged frontend files."""
    try:
        result = subprocess.run(
            ["git", "diff", "--cached", "--name-only"],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
        )
        if result.returncode != 0:
            return []

        output = result.stdout.strip()
        if not output:
            return []
        files = output.split("\n")
        frontend_files = [f for f in files if matches_frontend_pattern(f)]
        return frontend_files
    except Exception:
        return []


def get_modified_frontend_files() -> list[str]:
    """Get list of modified (unstaged) frontend files.

    Used to detect files that would be staged by `git commit -a`.
    """
    try:
        result = subprocess.run(
            ["git", "diff", "--name-only"],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
        )
        if result.returncode != 0:
            return []

        output = result.stdout.strip()
        if not output:
            return []
        files = output.split("\n")
        frontend_files = [f for f in files if matches_frontend_pattern(f)]
        return frontend_files
    except Exception:
        return []


def check_ui_verification_done(branch: str) -> bool:
    """Check if UI verification was confirmed for this branch."""
    safe_branch = sanitize_branch_name(branch)
    log_file = MARKERS_LOG_DIR / f"ui-check-{safe_branch}.done"
    return log_file.exists()


def main():
    """PreToolUse hook for git commit with frontend file changes."""
    result = {"decision": "approve"}

    try:
        data = parse_hook_input()
        tool_input = data.get("tool_input", {})
        command = tool_input.get("command", "")

        # Only check git commit commands
        if not is_git_commit_command(command):
            pass  # Early return case - still log at end
        else:
            # Check for staged frontend files
            frontend_files = get_staged_frontend_files()

            # Also check modified files if -a flag is used (git commit -a stages modified files)
            if has_auto_stage_flag(command):
                modified_files = get_modified_frontend_files()
                # Combine and deduplicate
                frontend_files = list(set(frontend_files + modified_files))

            if not frontend_files:
                pass  # No frontend files - still log at end
            else:
                # Get current branch
                branch = get_current_branch()
                if branch is None or branch in ("main", "master"):
                    pass  # Skip on main/master - still log at end
                elif check_ui_verification_done(branch):
                    pass  # Browser verification already confirmed - still log at end
                else:
                    # Block: frontend files staged but no browser verification confirmation
                    files_list = "\n".join(f"  - {f}" for f in sorted(frontend_files))
                    reason = (
                        f"フロントエンドファイルが変更されていますが、ブラウザ確認が完了していません。\n\n"
                        f"変更されたファイル:\n{files_list}\n\n"
                        "**必須手順**:\n"
                        "1. 開発サーバーを起動: `pnpm dev:frontend` (+ `pnpm dev:worker` if needed)\n"
                        "2. Chrome DevTools MCPで実際の動作を確認\n"
                        "   - `mcp__chrome-devtools__navigate_page` でアプリにアクセス\n"
                        "   - `mcp__chrome-devtools__take_snapshot` でDOM状態を確認\n"
                        "   - `mcp__chrome-devtools__list_console_messages` でエラーがないか確認\n"
                        "   - 必要に応じて `mcp__chrome-devtools__list_network_requests` でAPI/Analytics確認\n"
                        "3. 確認完了後、以下を実行:\n\n"
                        "```bash\n"
                        "python3 .claude/scripts/confirm-ui-check.py\n"
                        "```\n\n"
                        "その後、再度コミットを実行してください。"
                    )
                    result = make_block_result("ui-check-reminder", reason)

    except Exception as e:
        print(f"[ui-check-reminder] Hook error: {e}", file=sys.stderr)
        result = {"decision": "approve"}

    # Always log execution for accurate statistics
    log_hook_execution("ui-check-reminder", result.get("decision", "approve"), result.get("reason"))
    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/ui_check_reminder.ts
#!/usr/bin/env bun
/**
 * フロントエンド変更時にブラウザ確認を強制。
 *
 * Why:
 *   フロントエンドファイル変更後にブラウザ確認せずコミットすると、
 *   ランタイムエラーやUI崩れに気づかないまま本番に反映される。
 *
 * What:
 *   - git commit時（PreToolUse:Bash）に発火
 *   - frontend/src/配下のTS/TSX/JSON/CSS変更を検出
 *   - ブラウザ確認マーカーファイルがなければブロック
 *   - confirm_ui_check.ts実行後にコミット可能
 *
 * State:
 *   - reads: .claude/logs/markers/ui-check-*.done
 *
 * Remarks:
 *   - ブロック型フック（確認なしはコミット不可）
 *   - main/masterブランチはスキップ
 *   - .ts/.tsxはlib/hooks/workersも含む（Issue #209）
 *   - Python版: ui_check_reminder.py
 *
 * Changelog:
 *   - silenvx/dekita#xxx: フック追加
 *   - silenvx/dekita#209: .ts/.tsxファイル対象を拡大
 *   - silenvx/dekita#2917: TypeScript版初期実装
 *   - silenvx/dekita#2998: getMarkersDir()使用でworktree→メインリポジトリ解決
 *   - silenvx/dekita#3032: hasAutoStageFlag()の正規表現修正（--amend誤検出）
 */

import { execSync } from "node:child_process";
import { existsSync } from "node:fs";
import { TIMEOUT_MEDIUM } from "../lib/constants";
import { formatError } from "../lib/format_error";
import { getCurrentBranch } from "../lib/git";
import { logHookExecution } from "../lib/logging";
import { getMarkersDir } from "../lib/markers";
import { makeBlockResult } from "../lib/results";
import { parseHookInput } from "../lib/session";
import { sanitizeBranchName, stripQuotedStrings } from "../lib/strings";

const HOOK_NAME = "ui-check-reminder";

/**
 * File patterns that require browser verification.
 * Expanded to cover ALL frontend source files (see Issue #209 for background).
 */
const FRONTEND_FILE_PATTERNS = [
  "frontend/src/**/*.ts", // All TypeScript files (lib, hooks, workers, etc.)
  "frontend/src/**/*.tsx", // All React components
  "frontend/src/i18n/locales/*.json", // i18n translations
  "frontend/src/index.css", // Global CSS
];

/**
 * Check if a file path matches any frontend file pattern.
 */
export function matchesFrontendPattern(filepath: string): boolean {
  for (const pattern of FRONTEND_FILE_PATTERNS) {
    // fnmatch doesn't support **, so we need to handle it manually
    if (pattern.includes("**")) {
      // Split pattern at **
      const parts = pattern.split("**");
      if (parts.length === 2) {
        const prefix = parts[0];
        let suffix = parts[1];
        // Check if file starts with prefix and ends with suffix pattern
        if (filepath.startsWith(prefix)) {
          const remaining = filepath.slice(prefix.length);
          // suffix might have a leading /, remove it
          if (suffix.startsWith("/")) {
            suffix = suffix.slice(1);
          }
          // Simple glob match for suffix
          if (
            simpleGlobMatch(remaining, `*${suffix}`) ||
            simpleGlobMatch(remaining, `*/${suffix}`)
          ) {
            return true;
          }
        }
      }
    } else {
      if (simpleGlobMatch(filepath, pattern)) {
        return true;
      }
    }
  }
  return false;
}

/**
 * Simple glob matching (supports * wildcard).
 */
function simpleGlobMatch(str: string, pattern: string): boolean {
  // Escape special regex chars except *
  const regexPattern = pattern
    .split("*")
    .map((s) => s.replace(/[.+?^${}()|[\]\\]/g, "\\$&"))
    .join(".*");
  const regex = new RegExp(`^${regexPattern}$`);
  return regex.test(str);
}

/**
 * Check if command is a git commit command.
 */
export function isGitCommitCommand(command: string): boolean {
  if (!command.trim()) {
    return false;
  }
  const strippedCommand = stripQuotedStrings(command);
  return /git\s+commit\b/.test(strippedCommand);
}

/**
 * Check if command has -a or -am flag (auto-staging modified files).
 * Note: Only detects -a, -am, -ma, --all explicitly.
 * Combined flags like -av, -as are not detected (uncommon usage).
 */
export function hasAutoStageFlag(command: string): boolean {
  const strippedCommand = stripQuotedStrings(command);
  // Match git commit with -a, -am, -ma, or --all flags explicitly
  // Note: Previous regex was too greedy and matched --amend incorrectly
  // Use non-greedy .*? and require space before flag
  return /git\s+commit\b.*?\s(-a\b|-am\b|-ma\b|--all\b)/.test(strippedCommand);
}

/**
 * Get list of staged frontend files.
 */
function getStagedFrontendFiles(): string[] {
  try {
    const result = execSync("git diff --cached --name-only", {
      encoding: "utf-8",
      timeout: TIMEOUT_MEDIUM * 1000,
      stdio: ["pipe", "pipe", "pipe"],
    });

    const output = result.trim();
    if (!output) {
      return [];
    }
    const files = output.split("\n");
    return files.filter((f) => matchesFrontendPattern(f));
  } catch {
    return [];
  }
}

/**
 * Get list of modified (unstaged) frontend files.
 * Used to detect files that would be staged by `git commit -a`.
 */
function getModifiedFrontendFiles(): string[] {
  try {
    const result = execSync("git diff --name-only", {
      encoding: "utf-8",
      timeout: TIMEOUT_MEDIUM * 1000,
      stdio: ["pipe", "pipe", "pipe"],
    });

    const output = result.trim();
    if (!output) {
      return [];
    }
    const files = output.split("\n");
    return files.filter((f) => matchesFrontendPattern(f));
  } catch {
    return [];
  }
}

/**
 * Check if UI verification was confirmed for this branch.
 */
function checkUiVerificationDone(branch: string): boolean {
  const markersDir = getMarkersDir();
  const safeBranch = sanitizeBranchName(branch);
  const logFile = `${markersDir}/ui-check-${safeBranch}.done`;
  return existsSync(logFile);
}

async function main(): Promise<void> {
  let result: { decision?: string; reason?: string } = {};
  let sessionId: string | undefined;

  try {
    const data = await parseHookInput();
    sessionId = data.session_id;
    const toolInput = data.tool_input ?? {};
    const command = (toolInput.command as string) ?? "";

    // Only check git commit commands
    if (isGitCommitCommand(command)) {
      // Check for staged frontend files
      let frontendFiles = getStagedFrontendFiles();

      // Also check modified files if -a flag is used (git commit -a stages modified files)
      if (hasAutoStageFlag(command)) {
        const modifiedFiles = getModifiedFrontendFiles();
        // Combine and deduplicate
        const allFiles = new Set([...frontendFiles, ...modifiedFiles]);
        frontendFiles = Array.from(allFiles);
      }

      if (frontendFiles.length > 0) {
        // Get current branch
        const branch = await getCurrentBranch();
        if (branch !== null && branch !== "main" && branch !== "master") {
          if (!checkUiVerificationDone(branch)) {
            // Block: frontend files staged but no browser verification confirmation
            const filesList = frontendFiles
              .sort()
              .map((f) => `  - ${f}`)
              .join("\n");
            const reason = `フロントエンドファイルが変更されていますが、ブラウザ確認が完了していません。\n\n変更されたファイル:\n${filesList}\n\n**必須手順**:\n1. 開発サーバーを起動: \`pnpm dev:frontend\` (+ \`pnpm dev:worker\` if needed)\n2. Chrome DevTools MCPで実際の動作を確認\n   - \`mcp__chrome-devtools__navigate_page\` でアプリにアクセス\n   - \`mcp__chrome-devtools__take_snapshot\` でDOM状態を確認\n   - \`mcp__chrome-devtools__list_console_messages\` でエラーがないか確認\n   - 必要に応じて \`mcp__chrome-devtools__list_network_requests\` でAPI/Analytics確認\n3. 確認完了後、以下を実行:\n\n\`\`\`bash\nbun run .claude/scripts/confirm_ui_check.ts\n\`\`\`\n\nその後、再度コミットを実行してください。`;
            result = makeBlockResult(HOOK_NAME, reason);
          }
        }
      }
    }
  } catch (error) {
    console.error(`[${HOOK_NAME}] Hook error: ${formatError(error)}`);
    result = {};
  }

  // Always log execution for accurate statistics
  await logHookExecution(HOOK_NAME, result.decision ?? "approve", result.reason, undefined, {
    sessionId,
  });
  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main().catch((err) => {
    console.error(`[${HOOK_NAME}] Unhandled error:`, err);
    console.log(JSON.stringify({}));
    process.exit(1);
  });
}
